rust_src/src/fonts.rs

```diff
@@ -167,6 +169,48 @@ impl From<LispObject> for Option<LispFontObjectRef> {
     }
 }
 
+pub type LispFontSpecRef = ExternalPtr<Lisp_Font_Spec>;
+
+impl From<LispFontSpecRef> for LispObject {
+    fn from(f: LispFontSpecRef) -> Self {
+        LispObject::tag_ptr(f, Lisp_Type::Lisp_Vectorlike)
+    }
+}
+
+impl From<LispObject> for LispFontSpecRef {
+    fn from(o: LispObject) -> Self {
+        match o.into() {
+            Some(spec) => spec,
+            None => wrong_type!(Qfont_spec, o),
+        }
+    }
+}
+
+impl From<LispObject> for Option<LispFontSpecRef> {
+    fn from(o: LispObject) -> Self {
+        o.as_vectorlike().and_then(|v| {
+            if v.is_pseudovector(pvec_type::PVEC_FONT) && o.is_font_spec() {
+                Some(unsafe { mem::transmute(o) })
+            } else {
+                None
+            }
+        })
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn font_vconcat_entity_vectors(list: LispObject) -> LispObject {
+    vconcat_entity_vectors(list.into())
+}
+
+/// Concatenate all elements of LIST into one vector. LIST is a list of font-entity vectors.
+fn vconcat_entity_vectors(list: LispCons) -> LispObject {
```

Comment:
My only concern with using Rust's vectors here is SAFE_ALLOCA can trigger garbage collection at the end of the function. I do not believe this code will do that as written. It is _probably_ ok.